### ISSUE TITLE
Contribute Dockerfile ?

### ISSUE THREAD

**Issue by behrica** [vega/vega-lite#7801]
As I had a bit of an issue, to get the development environment for vega-lite working,
I made a Dockerfile.
For a complex setup such as vega-lite, with a lot of different build commands,
I was not really sure, what the Dockerfile "should do" as a command.

Same is true, for "mounting current workdir" vs "doing a git checkout".

But a least it gives a setup, in which all the different "yarn xxx" commands work.

I could contribute this Dockerfile, if you would find it usefull

**Comment by behrica** [comment]
This would be as well a step into supporting development environment as a Gitpod

**Comment by domoritz** [comment]
I'd be interested if you could also add automated tests and make sure the vega editor works in the docker environment as well. Having all of this work together would be great.

**Comment by behrica** [comment]
All tests pass "running inside the container". Not sure, if you mean that by automated tests.

**Comment by behrica** [comment]
I mean the Dockerfile I have gives you "a very quick way to start contributing" (without the hassle to setup node, yarn, jeckyl, bundler, ruby ) That's is done by the Dockerfile.
So after running 

```
git clone https://github.com/vega/vega-lite.git
cd vega-lite
docker build -t vega-lite-dev .
docker run -ti -p 4000:4000   vega-lite-dev
```

a potential contributor is in a bash shell, in which all "yarn" subcommands "just work".

**Comment by behrica** [comment]
I looked at the vega-editor here:
https://github.com/vega/editor
It seems to have it's own Dockerfile already.

Are you asking for a Dockerfile / image in which **both** can be done ?
* building vega-lite and running the vega lite site
* run the vega lite editor

**Comment by domoritz** [comment]
By automated tests I mean something in github that exercises the docker file just to make sure we are not breaking it. 

The editor is great for debugging so it would be nice to have it available in the docker as well. The tricky bit is that one needs to use 'yarn link' to use the dev version of Vega-Lite in the editor and I have not figured out how to make that work in docker. If you find a clean way, please add support for it. 

Also, make sure to update the contributing.md file in your pull request. 

Thank you.

**Comment by domoritz** [comment]
90% of the time I don't run the site as a developer btw.

**Comment by behrica** [comment]
> By automated tests I mean something in github that exercises the docker file just to make sure we are not breaking it.
> 
> The editor is great for debugging so it would be nice to have it available in the docker as well. The tricky bit is that one needs to use 'yarn link' to use the dev version of Vega-Lite in the editor and I have not figured out how to make that work in docker. If you find a clean way, please add support for it.

I am completely new to this, but the `yarn link` docu says:
--
Links are registered in ~/.config/yarn/link
--

This means that the folder  `~/.config/yarn` could be shared between 2 running containers (by mounting it as a volume into both) and then it should work.

Or alternatively both container mount the whole $home folder.
This can maybe "automagically" done, via a docker compose  file , which runs the  2 docker containers and configures this sharing.

**Comment by domoritz** [comment]
What do you think about a separate docker environment where the linking happens in the home directory of the docker container rather than the host system?

**Comment by behrica** [comment]
But this would require that vega editor and the "vega-lite" code (+ dev tools) would live in the same container.
This is probably not impossible, but against the Docker spirit of "one container per service"

"Development environments"  in Docker are stretching anyway the purpose of Docker.

The vega editor Dockerfile does correctly one thing only, it starts the editor on port 8080.
I think this should not be mixed with something else.

In my opinion the sharing of host directories is a well known practice in Docker, and considered a "clean" way to make certain type of application communicate.

Basically it would make 2 yarn installations / or 2 running yarn tools communicate with each other. It should not be different from having 2 yarn in two bashes on the same host communicating, given that permissions / owners are set correctly.
This is the tricky part. Both containers need to run the "yarn" command under the id of the host user.

**Comment by domoritz** [comment]
I see. We have to make sure that we don't mount the whole user directory, though, since that seems counter the spirit of only mounting what is really needed.

**Comment by behrica** [comment]
Yes. I am wondering up to which point the "linking" done via "yarn link" need to be taken into consideration from the beginning.

**Comment by behrica** [comment]
My "main use case" for the Dockerfile (and the reason I started it),
was to get "yarn site" working.
This requires quite some tools to setup, and I did not want to "spoil" my host OS with a plethora of Javascript / ruby stuff, which I never use outside vega lite.

This is of course only a subset of "functionality".
I could try to further test the setup, by making sure that all yarn commands specified in package.json run without errors.

If this is working, would you agree to review a Draft PR, containing the Dockerfile and continue based on that ?

**Comment by domoritz** [comment]
yes, I would be happy to review a pull request. My requirements for acceptance would be that the docker file is exercised ina github action and that there are basic instructions for using docker in the contributing docs.
